docs/rfcs/20210111_centralized_configuration_management.md

```diff
@@ -0,0 +1,588 @@
+prefix. Command line arguments passed to redpanda takes precedence over
+environment variables if both are set.
+
+Example configuration flags and corresponding environment variables:
+
+`--data-directory` `REDPANDA_DATA_DIRECTORY`
+`--heartbeat-interval-ms` `REDPANDA_HEARTBEAT_INTERVAL_MS`
+
+#### Collections
+
+For command line variables collections of parameters is set by passing
+in the same flag more than once.
+f.e:
+
+```bash
+redpanda --seed-server 127.0.0.1:1234 --seed-server 127.0.0.1:1234
+```
+
+This approach is impossible to use with environment variables as each
+assignment overrides previous value. To express collections 
+
+## Configuration property value
+
+Redpanda supports multiple sources for values of configuration properties.
+The configuration property precedence policy clearly defines what would be the
+final value of a configuration property if its value is different in each source.
+Redpanda defines a simple policy for configuration properties value precedence
+as property value is always equal to the one persisted in configuration storage.
+Configuration storage works according to the 'Last write wins' rule.
+This way, the most recent value set is used.
+When redpanda is started with set of flags or environment variables value of
+properties set with flags are persisted and stored values are overridden.
+If configuration change will be propagated to running redpanda instance later on
+it will override stored configuration value.
+This way redpanda always treat stored configuration as a source of truth.
+
+### Features of configuration properties
+
+Configuration properties may have different features.
+Redpanda configuration properties may be described by set of features.
+Those features determine policy of updates, persistance and programmatic API of
+the configuration properties.
+
+#### Scope
+
+Feature describing property range and the at the same time persistence policy.
+
+Possible values: `node/cluster`
+
+- `node` -  properties concern only single node
+- `cluster` - cluster properties will eventually have the same value on all the 
+              nodes, once set are propagated to all the nodes
+
+#### Change policy
+
+Possible values: `runtime/restart/invariant`
+
+- `invariant` - if property is invariant it can not be changed
+                without pruning node data, changing `invariant` property will
+                stop redpanda from starting
+
+- `runtime` - runtime changeable properties are the one that when updated are
+              immediately visible to all dependent components, they may change
+              in any point in time
+
+- `restart` - properties which require restart for the change to applied.
+              In future we may have less and less of this properties as we make
+              more features runtime configurable
+
+Example features of currently existing redpanda properties:
+
+`node_id` - local,invariant
+`developer_mode` - local,restart
```

Comment:
`developer_mode` is such an odd thing to deal with here. Perhaps we have an option that is `--dev` for developer mode that is mostly used by containers and such. If you still via rpm/deb package, then `--dev` mode is not passed by default?

Reply:
agreed. it seems like we could simply expose the things that developer mode does as their own flags or meta flags that turn on groups of similar options like developer mode with a name that is relevant. alternatively we could have some flags that hidden behind a `--enable-experimental` or something. in any case, i agree with the sentiment that developer mode flag could be improved.